### PR TITLE
Wait until stderr is consumed fully

### LINKED ISSUE #297: Unstable unit test
```
[ERROR] Failures: 
[ERROR]   ExternalTypescriptSensorTest.should_log_error_if_no_typescript_found:268 
Expecting:
 <["No TypeScript compiler found in your project",
    "Global one referenced in 'NODE_PATH' will be used",
    "Analyzing 1 typescript file(s) with the following configuration file /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/tsconfig.json",
    "Starting external process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts`",
    "Error: Cannot find module 'typescript'",
    "External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts` returned an empty output. Run with -X for more information"]>
to contain:
 <["Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project"]>
but could not find:
 <["Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project"]>
```

```
[INFO] Running org.sonar.plugin.typescript.ExternalTypescriptSensorTest
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockEmptyResponse.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/foo/file.ts` returned an empty output. Run with -X for more information
ERROR Some error message
ERROR No tsconfig.json file found for /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/file.ts (looking up the directories tree). This file will not be analyzed.
ERROR Failed to find input file for path `some/path/file.ts`
ERROR Error: Cannot find module 'typescript'
ERROR Failed to find 'typescript' module. Please check, NODE_PATH contains location of global 'typescript' or install locally in your project
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/mockTypescriptNotFound.js /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/src/test/resources/dirWithoutTypeScript/dirWithoutTypeScript/file.ts` returned an empty output. Run with -X for more information
ERROR Some error
ERROR 
ERROR /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classe...
ERROR throw errorMessage;
ERROR ^
ERROR External process `target/node/node /home/travis/build/SonarSource/SonarTS/sonarts-sq-plugin/sonar-typescript-plugin/target/test-classes/rulesDeadlock.js` returned an empty output. Run with -X for more information
[ERROR] Tests run: 13, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.009 s <<< FAILURE! - in org.sonar.plugin.typescript.ExternalTypescriptSensorTest
[ERROR] should_log_when_empty_sonarts_out(org.sonar.plugin.typescript.ExternalTypescriptSensorTest)  Time elapsed: 0.077 s  <<< ERROR!
java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at org.sonar.plugin.typescript.ExternalTypescriptSensorTest.should_log_when_empty_sonarts_out(ExternalTypescriptSensorTest.java:187)
ERROR abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabc...
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ExternalTypescriptSensorTest.should_log_when_empty_sonarts_out:187 » IndexOutOfBounds

```

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #297

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalProcessErrorConsumer.java [hunk 1]
```diff
@@ -39,9 +40,9 @@
 
   private static final Logger LOG = Loggers.get(ExternalProcessErrorConsumer.class);
 
-  public final void consumeStdError(Process process) {
+  public final Future<?> consumeStdError(Process process) {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalProcessErrorConsumer.java]: CRITICAL Remove usage of generic wildcard type. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1452)

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 2]
```diff
@@ -172,25 +175,37 @@ private static File findTsConfig(InputFile inputFile, File projectBaseDir) {
     ProcessBuilder processBuilder = new ProcessBuilder(command.commandLineTokens());
     setNodePath(localTypescript, processBuilder);
     LOG.debug(String.format("Starting external process `%s`", commandLine));
+    Process process = null;
     try {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: CRITICAL Change this "try" to a try-with-resources. (sonar.java.source not set. Assuming 7 or greater.) [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2093)

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 3]
```diff
@@ -49,6 +51,7 @@
 import org.sonar.api.measures.FileLinesContextFactory;
 import org.sonar.api.measures.Metric;
 import org.sonar.api.rule.RuleKey;
+import org.sonar.api.utils.command.CommandExecutor;
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MINOR Remove this unused import 'org.sonar.api.utils.command.CommandExecutor'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 4]
```diff
@@ -383,4 +381,24 @@ protected void readErrors(BufferedReader errorReader) {
...
+      running = true;
+      sleep();
+      super.readErrors(errorReader);
+      running = false;
+    }
+  }
+
+  private static void sleep() {
+    try {
+      Thread.sleep(500);
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 5]
```diff
@@ -383,4 +381,24 @@ protected void readErrors(BufferedReader errorReader) {
     }
   }
 
+  private static class SlowErrorConsumer extends ExternalProcessErrorConsumer {
```
Comment (inverno) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: @saberduck I would change the name of this one to `TestableErrorConsumer`, make the duration of sleep a constructor parameter and set it to `0` for all test cases but one.

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 6]
```diff
@@ -383,4 +384,29 @@ protected void readErrors(BufferedReader errorReader) {
...
+      running = true;
+      sleep(delay);
+      super.readErrors(errorReader);
+      running = false;
+    }
+  }
+
+  private static void sleep(int millis) {
+    try {
+      Thread.sleep(millis);
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)

**Pull request merged**